snapcraft/cli/_options.py

```diff
@@ -65,49 +65,112 @@ def prompt_for_value(self, ctx):
+    for option in reversed(options):
+        # Pop param_decls option to prevent exception further on down the
+        # line for: `got multiple values for keyword argument`.
+        option = option.copy()
+        param_decls = option.pop("param_decls")
+        click_option = click.option(param_decls, **option, hidden=hidden)
+        func = click_option(func)
+    return func
+
+
 def add_build_options(hidden=False):
     def _add_build_options(func):
-        for build_option in reversed(_BUILD_OPTIONS):
-            # Pop param_decls option to prevent exception further on down the
-            # line for: `got multiple values for keyword argument`.
-            build_option = build_option.copy()
-            param_decls = build_option.pop("param_decls")
-            option = click.option(param_decls, **build_option, hidden=hidden)
-            func = option(func)
-        return func
+        return _add_options(_BUILD_OPTIONS, func, hidden)
 
     return _add_build_options
 
 
-def get_build_environment(**kwargs):
-    force_use_lxd = kwargs.get("use_lxd")
-    force_destructive_mode = kwargs.get("destructive_mode")
+def add_provider_options(hidden=False):
+    def _add_provider_options(func):
+        return _add_options(_PROVIDER_OPTIONS, func, hidden)
 
-    if force_destructive_mode and force_use_lxd:
-        raise click.BadOptionUsage(
-            "--destructive-mode",
-            "--use-lxd and --destructive-mode cannot be used together.",
+    return _add_provider_options
+
+
+def get_build_environment(**kwargs) -> Tuple[str, bool]:
+    """Get build provider and determine if running as managed instance."""
+    provider = kwargs.get("provider")
+    use_lxd = kwargs.get("use_lxd")
+    destructive_mode = kwargs.get("destructive_mode")
+    env_provider = os.getenv("SNAPCRAFT_BUILD_ENVIRONMENT")
+
+    if destructive_mode and use_lxd:
+        raise click.BadArgumentUsage(
+            "--use-lxd and --destructive-mode cannot be used together."
         )
-    elif force_use_lxd:
-        provider = "lxd"
-    elif force_destructive_mode:
+
+    if provider:
+        # Sanity check for mismatch between provider flag and env.
+        if env_provider and env_provider != provider:
+            raise click.BadArgumentUsage(
+                "mismatch between --provider={} and SNAPCRAFT_BUILD_ENVIRONMENT={}".format(
+                    provider, env_provider
+                )
+            )
+
+        if provider != "lxd" and use_lxd:
+            raise click.BadArgumentUsage(
+                "--use-lxd cannot be used with --provider={}".format(provider)
+            )
+
+        if provider != "host" and destructive_mode:
+            raise click.BadArgumentUsage(
+                "--destructive-mode cannot be used with --provider={}".format(provider)
+            )
+
+        # Specifying --provider=host requires the use of --destructive-mode.
+        # SNAPCRAFT_BUILD_ENVIRONMENT=host does not.
```

Comment:
maybe it should, our tests would need to adapt and we would need an MP for launchpad to use --destructive-mode